crates/voicevox_core_c_api/src/compatible_engine.rs

```diff
@@ -359,6 +359,112 @@ pub unsafe extern "C" fn decode_forward(
     }
 }
 
+/// # Safety
+///
+/// - `f0`はRustの`&[f32; length as usize]`として解釈できなければならない。
+/// - `phoneme`はRustの`&[f32; phoneme_size * length as usize]`として解釈できなければならない。
+/// - `speaker_id`はRustの`&[i64; 1]`として解釈できなければならない。
+/// - `output`はRustの`&mut [MaybeUninit<f32>; ((length + 2 * margin_width) * feature_dim) as usize]`として解釈できなければならない。
+#[unsafe(no_mangle)] // SAFETY: voicevox_core_c_apiを構成するライブラリの中に、これと同名のシンボルは存在しない
+pub unsafe extern "C" fn generate_full_intermediate(
+    length: i64,
+    phoneme_size: i64,
+    margin_width: i64,
+    feature_dim: i64,
+    f0: *mut f32,
+    phoneme: *mut f32,
+    speaker_id: *mut i64,
+    output: *mut f32,
```

Comment:
特に規則は決まっていないのですが、入力引数は推論に必要なパラメータだけになっていそうです。
`generate_full_intermediate`は`decode`と同じシグネチャになってる方が良さそう！
（`margin_width`は`generate_full_intermediate`内でMARGINが必要なので、後々のことを考えると要りそう）

```suggestion
pub unsafe extern "C" fn generate_full_intermediate(
    length: i64,
    phoneme_size: i64,
    margin_width: i64,
    f0: *mut f32,
    phoneme: *mut f32,
    speaker_id: *mut i64,
    output: *mut f32,
```

Reply:
出力の配列サイズのチェックをするなら、output_lengthを追加することで、渡した*outputの確保サイズを教えるという形がいいのかなと思いました。`decode`のシグネチャには入っていませんが...

Reply:
> 出力の配列サイズのチェックをするなら、output_lengthを追加することで、渡した*outputの確保サイズを教えるという形がいいのかなと思いました。`decode`のシグネチャには入っていませんが...

よくわかっていないのですが、RustでアロケートするわけではなくNumPyの`buffer`に書き込むはずなので上手くいかない気がします。
（やるならPythonの世界に対して二段階（サイズ→中身）に情報を伝達する仕組みが必要なんじゃないかと）

Reply:
あ、outputのバリデーションはなしが良いと思っています！
やったほうが良いかもですが、compatible engineの他のAPIではやってないので揃えるほうが重要だという判断です 🙇

`generate_full_intermediate`にとって`margin_width`は確かに不要な気がしました！！

まあぶっちゃけこれ何を引数に与えるべきかとかどういう設計にすべきなのかとかそこまでよく考えずに作られたシグネチャなので、とりあえず今動けば何でも良い気がしています 😇 
たぶん「将来値が変わってもそのまま動作するようにするため」とかだと思うのですが、なんか設計が中途半端な気がしていて、あまり考えずにそれっぽい引数にするので良いのかなと。。 😇

Reply:
あ、誤読してたかも。`output_length`ってENGINE→COREの通達ですかね？

---

> なんか設計が中途半端な気がしていて、あまり考えずにそれっぽい引数にするので良いのかなと。。 😇 

同意ですね… VOICEVOX ENGINEしか使わないので、後で困る可能性があるか否かだけを考えればよさそう。それでmarginはどちらかというとCORE側が知っているはずなので、[`14`から変えるときになってからCORE→ENGINEに別途伝達する経路を用意すればよさそう](https://github.com/VOICEVOX/voicevox_core/issues/866#issuecomment-2457854468)。